compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1173,8 +1167,19 @@ private boolean isXMLTypeAssignable(BType sourceType, BType targetType, Set<Type
                 }
                 return isAssignable(source.constraint, targetType, unresolvedTypes);
             }
-        } else if (sourceTag == TypeTags.XML_TEXT && targetTag == TypeTags.STRING) {
-            return true;
+        } else if (sourceTag == TypeTags.XML_TEXT) {
+            if (targetTag == TypeTags.STRING || targetTag == TypeTags.CHAR_STRING) {
+                return true;
+            }
+
+            if (targetTag == TypeTags.FINITE) {
+                BFiniteType finiteType = (BFiniteType) targetType;
+                for (BLangExpression finiteValue : finiteType.getValueSpace()) {
+                    if (isXMLTypeAssignable(sourceType, finiteValue.type, unresolvedTypes)) {
```

Comment:
Checking by type will not work, right? The following compiles with these changes, but should result in a compilation error?

```ballerina
type XmlValue 1|"foo";

public function main() {
    xml:Text notFoo = xml `not foo`;
    XmlValue xmlVal = notFoo;
}
```

Reply:
We also need to introduce these changes to the runtime type checker, right?

Reply:
Actually, we probably don't need this check at all at compile time, right? We only need it at runtime?

Reply:
I think this gets invalidated when we remove the implicity conversion from xml text to string

Reply:
Allowing this was a mistake. Will remove this, thank you.